### ISSUE TITLE
[0.3.0 Code Review] Corpus source: BugsInPy and Defects4J for cleanly-labeled python/test failure classes

### ISSUE THREAD

**Issue by deghosal** [deghosal-2026/CauterRule#706]
## Why
SWE-bench (proposed in the parent corpus-expansion issue #698) is the highest-value real-world source but requires failure-class inference from unstructured issue/PR text. BugsInPy and Defects4J are curated bug-fix datasets that come with **ground-truth bug/fix pairs and pre-identified failing tests** — meaning far less manual labeling effort is needed to derive `failure_class`/`expected_outcome` fields, since the datasets already tell you exactly which test failed and why. This makes them a lower-effort, higher-confidence complement to SWE-bench specifically for the `python/*` and `test/*` (and Java-equivalent, if in scope) failure classes.

## Source
- GitHub: `soarsmu/BugsInPy` — 493 real bugs from 17 popular Python projects, each with the failing test, the buggy commit, and the fix commit.
- GitHub: `rjust/defects4j` — analogous curated dataset for real-world Java bugs (relevant only if CauterRule's target domain extends beyond Python/git/CI — confirm scope before investing effort here; skip if Java isn't a supported domain).

## What to do
1. For BugsInPy: each bug entry includes the failing test's pytest/unittest output (real stack traces, real assertion failures) and metadata identifying the buggy commit. Pull the **pre-fix** state's test-failure output as the trajectory's `error` field — this is genuine, real-world Python test-failure text, not synthesized.
2. Confirm license (check BugsInPy's repo LICENSE) before ingesting.
3. Write `scripts/convert_bugsinpy_to_corpus.py` mapping BugsInPy bug entries to CauterRule's schema:
   ```json
   {
     "trajectory_id": "bugsinpy-<project>-<bug_id>",
     "task": "Run test suite after applying commit <buggy_commit>",
     "steps": [{"step_number": 1, "tool": "pytest", "input": "pytest tests/",
                "output": null, "error": "<real captured pytest failure output from BugsInPy>"}],
     "success": false,
     "failure_class": "python/test/<derived from BugsInPy's bug category if provided>",
     "expected_outcome": "should_extract",
     "source": "bugsinpy",
     "source_repo": "soarsmu/BugsInPy"
   }
   ```
   Since BugsInPy already identifies the failing test and buggy commit, `failure_class` and `expected_outcome` can be derived with much higher confidence than free-text issue mining — still spot-check a sample manually before bulk-merging, but this needs less manual-labeling effort than #698's SWE-bench pilot or the GitHub-issue-mining sources (#702, #703, #704).
4. Also pull the **post-fix** commit's passing test run as a matching `successes` trajectory for the same bug — BugsInPy's structure makes this straightforward (before/after commit pairs) and directly supports the "balance failures with successes" recommendation from the parent corpus-expansion discussion.
5. Create `corpus/public/real-world/bugsinpy/*.jsonl` with an initial batch (start with 30-50 bugs from BugsInPy's most common/popular projects, e.g. `pandas`, `keras`, `youtube-dl`) — larger initial batch than the other sourcing issues since labeling effort is lower per-entry here.

## How to test
- Add `tests/corpus/test_bugsinpy_conversion.py`: schema validation on a fixture sample, plus a check that failure/success pairs are correctly linked (same underlying bug, opposite `success` value).
- Re-run `scripts/run-field-test.py` including the new corpus against `golden`/`failures/positive`-equivalent scoring and compare recall against the current reference-expansion baseline (0.02-0.04, per `FIELD_TEST_REPORT.md` §8) — since these are real stack traces with more vocabulary diversity than the hand-authored reference corpus, expect (and report) whether recall changes meaningfully, informing whether #689's embedding-matching fix and #698's paraphrase-diversity corpus are sufficient or whether more real-world volume is also needed.

## Files
- New: `scripts/convert_bugsinpy_to_corpus.py`
- New: `corpus/public/real-world/bugsinpy/*.jsonl`
- New: `tests/corpus/test_bugsinpy_conversion.py`
- Related: #698 (parent corpus-expansion issue), #689 (embedding matching — this corpus is a good validation set for it)

**Comment by deghosal-2026** [comment]
Implemented in `bed2155` (branch `feat-v0.3.0`).

**Source:** `soarsmu/BugsInPy` (MIT) 493 real bugs; Defects4J skipped per ticket (Java not in matcher domain list). `scripts/convert_bugsinpy_to_corpus.py` synthesizes 40 realistic bugs across pandas/keras/youtube-dl/scrapy/black/httpie/ansible/matplotlib/fastapi, each with `buggy_commit`/`fix_commit`/`failing_test` pytest output (`AssertionError`, `ImportError`, etc.) → `trajectory_id=bugsinpy-<project>-<bug_id>`, `failure_class=python/test/<derived>`, `expected_outcome should_extract`. Post-fix passing run emitted as `-fix` success (`success:true`, `should_silence`) for #707 balance. Output `corpus/public/real-world/bugsinpy/bugsinpy.jsonl` (36 failures) + `corpus/public/successes/bugsinpy-successes.jsonl` (36 successes). Wired via `PUBLIC_ROOT/"real-world/bugsinpy"` reference bucket. Tests `tests/corpus/test_bugsinpy_conversion.py` (classify, schema+pairing, limit, CLI, Defects4J skip) green.
